internal/jobobject/queue.go

```diff
@@ -0,0 +1,96 @@
+package jobobject
+
+import (
+	"errors"
+	"sync"
+)
+
+var (
+	errQueueClosed = errors.New("the queue is closed for reading and writing")
+	errQueueEmpty  = errors.New("the queue is empty")
+)
+
+// messageQueue represents a threadsafe message queue to be used to retrieve or
+// write notifications to. This will be used to hold and retrieve messages for job
+// objects and doesn't have much use in other packages. There should only be one reader
+// and one writer.
+type messageQueue struct {
+	m        sync.Mutex
+	messages []interface{}
+	ready    chan struct{} // Channel to signal that a message is available to read.
+	closed   bool
+}
+
+func newMessageQueue() *messageQueue {
+	return &messageQueue{
+		messages: []interface{}{},
+		ready:    make(chan struct{}),
+	}
+}
+
+func (mq *messageQueue) write(msg interface{}) error {
+	mq.m.Lock()
+	defer mq.m.Unlock()
+
+	if mq.closed {
+		return errQueueClosed
+	}
+	length := len(mq.messages)
+	mq.messages = append(mq.messages, msg)
+	// If the queue was empty, we need to signal that there's now a value ready to read.
+	if length == 0 {
+		// This handles the very first write case. If this is the first write, we could end
+		// up in a spot where we'd signal mq.ready but no one is waiting for a value so we'll block
+		// forever. This way we can add to the queue, so any call to read after will just return the
+		// value and any call to readOrWait should be listening for a new value and we'll hit the
+		// `mq.ready <- struct{}{}` case.
+		select {
+		case mq.ready <- struct{}{}:
+		default:
+		}
+	}
+	return nil
+}
+
+// read will read a value from the queue if available, otherwise return an error.
+func (mq *messageQueue) read() (interface{}, error) {
+	mq.m.Lock()
+	defer mq.m.Unlock()
+	if mq.closed {
+		return nil, errQueueClosed
+	}
+	if len(mq.messages) == 0 {
+		return nil, errQueueEmpty
+	}
+	val := mq.messages[0]
+	mq.messages[0] = nil
+	mq.messages = mq.messages[1:]
+	return val, nil
+}
+
+// readOrWait will read a value from the queue if available, else it will wait for a
+// value to become available. This will block forever if nothing gets written or until
+// the queue gets closed.
+func (mq *messageQueue) readOrWait() (interface{}, error) {
+	msg, err := mq.read()
+	if err == errQueueEmpty {
+		<-mq.ready
```

Comment:
Looks like there is a racey deadlock here. Consider this case:
- Goroutine 1 executes the `read` on line 75
- Goroutine 2 does a full `write` call
- Goroutine 1 continues executing with lines 76 and 77, but hangs on 77 because it already missed the `ready` send from the write

Reply:
Yeah that's a good point, because from the writers perspective, there would be no one yet reading the ready messages and thus it would take the default case of the select statement used. Then in here we would hang forever waiting on the ready message.